bundles/org.eclipse.e4.ui.workbench.addons.swt/src/org/eclipse/e4/ui/workbench/addons/cleanupaddon/CleanupAddon.java

```diff
@@ -381,4 +384,32 @@ private void subscribeRenderingChanged(
 	boolean isLastEditorStack(MUIElement element) {
 		return modelService.isLastEditorStack(element);
 	}
+
+	private void transferPrimaryDataStackIfRemoved(MElementContainer<?> removedElement) {
+		if (!PRIMARY_DATA_STACK_ID.equals(removedElement.getElementId())) {
+			return;
+		}
+		MPartStack newPrimaryDataStack = findFirstPartStackToBeRendered(removedElement.getParent());
+		if (newPrimaryDataStack != null) {
+			newPrimaryDataStack.setElementId(removedElement.getElementId());
+			newPrimaryDataStack.getTags().addAll(removedElement.getTags());
```

Comment:
I wonder, are there specific tags that you expect to be transferred?  Is it ever possible that there might be tags you'd not want transferred?

Reply:
I only expect the "EditorStack" and the "primary data stack" tags to be transferred. I think you are right that we should not simply transfer every kind of tag but only those related to the stack being the primary data stack. I have adapted the code accordingly with the introduced utility for adding the relevant tags as mentioned in the comment above (https://github.com/eclipse-platform/eclipse.platform.ui/pull/1775#discussion_r1542729945).